library/alloc/tests/vec.rs

```diff
@@ -1137,21 +1145,17 @@ fn test_from_iter_specialization_panic_during_drop_leaks() {
         }
     }
 
-    let mut to_free: *mut Droppable = core::ptr::null_mut();
-    let mut cap = 0;
-
     let _ = std::panic::catch_unwind(AssertUnwindSafe(|| {
-        let mut v = vec![Droppable::DroppedTwice(Box::new(123)), Droppable::PanicOnDrop];
-        to_free = v.as_mut_ptr();
-        cap = v.capacity();
-        let _ = v.into_iter().take(0).collect::<Vec<_>>();
+        let v = vec![
+            Droppable::ShouldBeDropped,
+            Droppable::DroppedTwice(Box::new(123)),
+            Droppable::PanicOnDrop,
+        ];
+        let _ = v.into_iter().take(1).collect::<Vec<_>>();
```

Comment:
maybe add a map() here that only creates a `ShouldBeDropped` variant when things logically get moved into the output vec?

The New/Old distinction from #101628 is helpful.